From 6dd0a4dc29890d6a0e7b2bcce851fbade35534cf Mon Sep 17 00:00:00 2001
From: Sergio Lopez <slp@redhat.com>
Date: Wed, 11 Nov 2020 07:00:32 +0100
Subject: [PATCH 9/9] krunfw: Allow RST if sk_shutdown == SHUTDOWN_MASK

Even if sk_shutdown == SHUTDOWN_MASK, allow processing of OP_RST
packets. This prevents RST storms caused by dangling sockets.

Signed-off-by: Sergio Lopez <slp@redhat.com>
---
 net/vmw_vsock/virtio_transport_common.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/net/vmw_vsock/virtio_transport_common.c b/net/vmw_vsock/virtio_transport_common.c
index ce6144f10e28..8a084c7d2c42 100644
--- a/net/vmw_vsock/virtio_transport_common.c
+++ b/net/vmw_vsock/virtio_transport_common.c
@@ -1282,7 +1282,9 @@ void virtio_transport_recv_pkt(struct virtio_transport *t,
 	lock_sock(sk);
 
 	/* Check if sk has been released before lock_sock */
-	if (sk->sk_shutdown == SHUTDOWN_MASK) {
+	if (sk->sk_shutdown == SHUTDOWN_MASK &&
+	    (sk->sk_state != TCP_ESTABLISHED ||
+	     le16_to_cpu(pkt->hdr.op) != VIRTIO_VSOCK_OP_RST)) {
 		(void)virtio_transport_reset_no_sock(t, pkt);
 		release_sock(sk);
 		sock_put(sk);
-- 
2.28.0

